File: aspnet-in-practice/14/CS/Cache/Cache/Cache/AppFabric/AppFabricOutputCacheProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Runtime.Caching;
using Cache;
using Microsoft.ApplicationServer.Caching;
using System.Web.Caching;

namespace Cache.Providers.AppFabric
{
	public class AppFabricOutputCacheProvider : OutputCacheProvider
	{
		static AppFabricCacheProvider provider = new AppFabricCacheProvider();

		public override object Add(string key, object entry, DateTime utcExpiry)
		{
			Set(key, entry, utcExpiry);			
			return entry;
		}

		public override object Get(string key)
		{
			return provider.Get(key);
		}

		public override void Remove(string key)
		{
			provider.Remove(key);
		}

		public override void Set(string key, object entry, DateTime utcExpiry)
		{
			provider.Set(key, entry, new CacheItemPolicy { AbsoluteExpiration = utcExpiry.ToLocalTime() });
		}
	}
}

File: aspnet-in-practice/14/VB/Cache/Cache/Providers/AppFabric/AppFabricOutputCacheProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Runtime.Caching
Imports Cache
Imports Microsoft.ApplicationServer.Caching
Imports System.Web.Caching

Namespace Providers.AppFabric
	Public Class AppFabricOutputCacheProvider
		Inherits OutputCacheProvider
		Shared provider As New AppFabricCacheProvider()

		Public Overrides Function Add(ByVal key As String, ByVal entry As Object, ByVal utcExpiry As DateTime) As Object
			[Set](key, entry, utcExpiry)
			Return entry
		End Function

		Public Overrides Function [Get](ByVal key As String) As Object
			Return provider.[Get](key)
		End Function

		Public Overrides Sub Remove(ByVal key As String)
			provider.Remove(key)
		End Sub

		Public Overrides Sub [Set](ByVal key As String, ByVal entry As Object, ByVal utcExpiry As DateTime)
			provider.[Set](key, entry, New CacheItemPolicy() With {
			 .AbsoluteExpiration = utcExpiry.ToLocalTime()
			})
		End Sub
	End Class
End Namespace